File: gui/src/main/java/com/github/fernthedev/pi_mp3/ui/UIFactory.java
package com.github.fernthedev.pi_mp3.ui;

public class UIFactory {

    private static UIInterface uiInterface;

    public static void setUiInterface(UIInterface uiInterface) {
        if (UIFactory.uiInterface != null) throw new IllegalStateException("UI Interface is already registered");

        UIFactory.uiInterface = uiInterface;
    }

    public static UIInterface getUiInterface() {
        return uiInterface;
    }
}

File: api/src/main/kotlin/com/github/fernthedev/pi_mp3/api/ui/UIFactory.kt
package com.github.fernthedev.pi_mp3.api.ui

interface UIFactory {
    /**
     * Creates the button
     * @param width
     * @param height
     * @return The button
     */
    fun createButton(position: Position, width: Double, height: Double): UIButton

    /**
     * Creates the button
     * The width and height are defined
     * by the implementation of the
     * [UIFactory]
     *
     * @return The button
     */
    fun createButton(position: Position): UIButton

    /**
     * Creates the button
     * @param width
     * @param height
     * @return The button
     */
    fun createTextbox(position: Position?, width: Double, height: Double): UITextbox

    /**
     * Creates the button
     * The width and height are defined
     * by the implementation of the
     * [UIFactory]
     *
     * @return The button
     */
    fun createTextbox(position: Position?): UITextbox
}